planner/core/rule_partition_processor.go

```diff
@@ -1575,13 +1576,32 @@ func (p *rangeColumnsPruner) partitionRangeForExpr(sctx sessionctx.Context, expr
 		return 0, len(p.data), false
 	}
 
+	exprCollation, err := expression.CheckAndDeriveCollationFromExprs(sctx, opName, col.RetType.EvalType(), con, p.data[0])
+	if err != nil {
+		return 0, len(p.data), false
+	}
+	if exprCollation.Coer < expression.CoercibilityImplicit {
+		// Explicit collation, if set to Binary collation, we can still use prune partitions on EQ,
+		// but we need to change the constants collation
+		conCharSet, conColl := con.CharsetAndCollation()
+		colCharSet, colColl := p.partCol.RetType.Charset, p.partCol.RetType.Collate
+		if !collate.IsBinCollation(conColl) || conCharSet != colCharSet || opName != ast.EQ {
+			return 0, len(p.data), true
+		}
+		clonedCon := con.Clone()
+		if newCon, ok := clonedCon.(*expression.Constant); ok {
+			con = newCon
+		}
+		con.SetCharsetAndCollation(conCharSet, colColl)
+	}
 	start, end := p.pruneUseBinarySearch(sctx, opName, con, op)
 	return start, end, true
 }
 
 func (p *rangeColumnsPruner) pruneUseBinarySearch(sctx sessionctx.Context, op string, data *expression.Constant, f *expression.ScalarFunction) (start int, end int) {
 	var err error
 	var isNull bool
+	charSet, collation := p.partCol.RetType.Charset, p.partCol.RetType.Collate
```

Comment:
the parameter `f` is not needed anymore after this line

Reply:
Thanks, fixed